pkg/plugins/policies/meshaccesslog/plugin/v1alpha1/plugin.go

```diff
@@ -24,7 +29,146 @@ func (p plugin) MatchedPolicies(dataplane *core_mesh.DataplaneResource, resource
 	return matchers.MatchedPolicies(api.MeshAccessLogType, dataplane, resources)
 }
 
+type listeners struct {
+	inbound  map[xds.InboundListener]*envoy_listener.Listener
+	outbound map[mesh_proto.OutboundInterface]*envoy_listener.Listener
+}
+
+func gatherListeners(rs *core_xds.ResourceSet) listeners {
+	inboundListeners := map[xds.InboundListener]*envoy_listener.Listener{}
+	outboundListeners := map[mesh_proto.OutboundInterface]*envoy_listener.Listener{}
+
+	for _, res := range rs.Resources(envoy_resource.ListenerType) {
+		listener := res.Resource.(*envoy_listener.Listener)
+		address := listener.GetAddress().GetSocketAddress()
+
+		switch res.Origin {
+		case generator.OriginOutbound:
+			outboundListeners[mesh_proto.OutboundInterface{
+				DataplaneIP:   address.GetAddress(),
+				DataplanePort: address.GetPortValue(),
+			}] = listener
+		case generator.OriginInbound:
+			inboundListeners[xds.InboundListener{
+				Address: address.GetAddress(),
+				Port:    address.GetPortValue(),
+			}] = listener
+		default:
+			continue
+		}
+	}
+
+	return listeners{
+		outbound: outboundListeners,
+		inbound:  inboundListeners,
+	}
+}
+
+func (p plugin) applyToInbounds(
+	policies xds.TypedMatchingPolicies, inboundListeners map[xds.InboundListener]*envoy_listener.Listener, dataplane *core_mesh.DataplaneResource,
+) error {
+	for _, inbound := range dataplane.Spec.GetNetworking().GetInbound() {
+		iface := dataplane.Spec.Networking.ToInboundInterface(inbound)
+
+		listenerKey := xds.InboundListener{
+			Address: iface.DataplaneIP,
+			Port:    iface.DataplanePort,
+		}
+		listener, ok := inboundListeners[listenerKey]
+		if !ok {
+			continue
+		}
+
+		serviceName := inbound.GetTags()[mesh_proto.ServiceTag]
+
+		var conf *api.MeshAccessLog_Conf
+		if computed := policies.FromRules.Rules[listenerKey].Compute(xds.Subset{{
+			Key: mesh_proto.ServiceTag, Value: serviceName,
+		}}); computed != nil {
+			conf = computed.(*api.MeshAccessLog_Conf)
+		} else {
+			continue
+		}
```

Comment:
I find this a little hard to read TBH.

```
conf = policies.FromRules.Rules[listenerKey].Compute(xds.Subset{{
			Key: mesh_proto.ServiceTag, Value: serviceName,
		}}).(*api.MeshAccessLog_Conf)
```
is not valid in the return is nil?

Maybe a `xds.NewServiceSubset(serviceName)` function would make this more readable?

You'd have:
```go
computed := policies.FromRules.Rules[listenerKey].Compute(xds.NewServiceSubset(serviceName))
if computed == nil {
  continue
}
conf := computed.(*api.MeshAccessLog_Conf)
```
Or maybe even cast in the `range`

Reply:
Typed nil, you end up with something like:

```
panic: interface conversion: proto.Message is nil, not *api.MeshAccessLog_Conf
```

I'm partial to the "assign inside the if" because it doesn't introduce anything extra into the scope. I introduced `xds.MeshService` to create a `MeshService`-like `xds.Subset`

Reply:
it might be a "nice to have" to generate monomorphized versions of all these `proto.Message` functions for each policy type to avoid casting everywhere